cores/arduino/UARTClass.cpp

```diff
@@ -168,7 +168,7 @@ int UARTClass::read( void )
 
 void UARTClass::flush( void )
 {
-  while (_tx_buffer->_iHead != _tx_buffer->_iTail); //wait for transmit data to be sent
+  while (_tx_buffer->_iHead != *(volatile int*)&(_tx_buffer->_iTail)); //wait for transmit data to be sent
```

Comment:
The reason is that iTail can be updated in the irq handler. 
Without the volatile, the compiler may optimize the code by reading it only once from memory, and keeping its value in register. That would lead to an infinite loop.

I don't remember if i actually ran into this problem or not, and i did not check the resulting assembly, but that look like a typical case where volatile is needed.

Note that this more related to pull request https://github.com/01org/corelibs-arduino101/pull/266 than to https://github.com/01org/corelibs-arduino101/pull/274